### PR TITLE
[ISSUE #6574] fix DefaultMQProducerImpl.request sync invoke timeout

### PR DESCRIPTION
**Pull request by haiyanghan**
**Make sure set the target branch to `develop`**

## What is the purpose of the change

fix #<xxx> 

## Brief changelog

XX

## Verifying this change

XXXX

Follow this checklist to help us incorporate your contribution quickly and easily. Notice, `it would be helpful if you could finish the following 5 checklist(the last one is not necessary)before request the community to review your PR`.

- [x] Make sure there is a [Github issue](https://github.com/apache/rocketmq/issues) filed for the change (usually before you start working on it). Trivial changes like typos do not require a Github issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue. 
- [x] Format the pull request title like `[ISSUE #123] Fix UnknownException when host config not exist`. Each commit in the pull request should have a meaningful subject line and body.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [x] Write necessary unit-test(over 80% coverage) to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add integration-test in [test module](https://github.com/apache/rocketmq/tree/master/test).
- [x] Run `mvn -B clean apache-rat:check findbugs:findbugs checkstyle:checkstyle` to make sure basic checks pass. Run `mvn clean install -DskipITs` to make sure unit-test pass. Run `mvn clean test-compile failsafe:integration-test`  to make sure integration-test pass.
- [ ] If this contribution is large, please file an [Apache Individual Contributor License Agreement](http://www.apache.org/licenses/#clas).

### CODE REVIEW

**Review by RongtongJin — DISMISSED** [review]

File: client/src/main/java/org/apache/rocketmq/client/impl/producer/DefaultMQProducerImpl.java [hunk 1]
```diff
@@ -1462,25 +1458,23 @@ public Message request(final Message msg, final MessageQueueSelector selector, f
...
                 public void onException(Throwable e) {
-                    requestResponseFuture.setSendRequestOk(false);
-                    requestResponseFuture.putResponseMessage(null);
-                    requestResponseFuture.setCause(e);
+                    requestFail(correlationId, e);
                 }
             }, timeout - cost);
 
-            return waitResponse(msg, timeout, requestResponseFuture, cost);
+            return waitResponse(msg, timeout - cost, responseFuture, cost);
```
Comment (cserwen) [inline client/src/main/java/org/apache/rocketmq/client/impl/producer/DefaultMQProducerImpl.java:1477]: `timeout` or `timeout - cost` ?

Comment (cserwen) [reply client/src/main/java/org/apache/rocketmq/client/impl/producer/DefaultMQProducerImpl.java:1477]: > time - cost More accurate, cost is pre handler time

+1

File: client/src/main/java/org/apache/rocketmq/client/impl/producer/DefaultMQProducerImpl.java [hunk 2]
```diff
@@ -1563,34 +1554,38 @@ public void request(final Message msg, final MessageQueue mq, final RequestCallb
...
-        RequestFutureHolder.getInstance().getRequestFutureTable().put(correlationId, requestResponseFuture);
+        RequestFutureHolder.getInstance().getRequestFutureTable().put(correlationId,
+                new RequestResponseFuture(correlationId, timeout, requestCallback));
 
         long cost = System.currentTimeMillis() - beginTimestamp;
         this.sendKernelImpl(msg, mq, CommunicationMode.ASYNC, new SendCallback() {
             @Override
             public void onSuccess(SendResult sendResult) {
-                requestResponseFuture.setSendRequestOk(true);
+                requestSuccess(correlationId, new Message(msg.getTopic(), msg.getBody()));
```
Comment (francisoliverlee) [inline client/src/main/java/org/apache/rocketmq/client/impl/producer/DefaultMQProducerImpl.java:1564]: how about message's properties ?

File: client/src/main/java/org/apache/rocketmq/client/impl/producer/DefaultMQProducerImpl.java (hunk 1 above)
Comment (haiyanghan) [reply client/src/main/java/org/apache/rocketmq/client/impl/producer/DefaultMQProducerImpl.java:1477]: enen, but the scope of current PR modification is the request method

**Pull request closed without merging**